In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid')
% matplotlib inline

### GATHERING DATA

#### IMAGE PREDICTIONS

In [2]:
import requests
from PIL import Image
from io import BytesIO
r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

In [3]:
# Save .tsv to file.
with open("image_predictions.tsv", mode = 'wb') as file:
    file.write(r.content)

In [4]:
# Read file and view first few rows.
image_predictions = pd.read_csv('image_predictions.tsv', sep = '\t')
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
image_predictions.shape

(2075, 12)

In [6]:
# Find count of true positives and true negatives for p1
p1_true = image_predictions.query('p1_dog == True').count()[0]
p1_false = image_predictions.query('p1_dog == False').count()[0]
p1_true, p1_false

(1532, 543)

In [7]:
# Find count of true positives and true negatives for p2
p2_true = image_predictions.query('p2_dog == True').count()[0]
p2_false = image_predictions.query('p2_dog == False').count()[0]
p2_true, p2_false

(1553, 522)

In [8]:
# Find count of true positives and true negatives for p3
p3_true = image_predictions.query('p3_dog == True').count()[0]
p3_false = image_predictions.query('p3_dog == False').count()[0]
p3_true, p3_false

(1499, 576)

#### TWITTER ARCHIVE

In [9]:
# Read twitter archive file and view first few rows
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [10]:
twitter_archive.shape

(2356, 17)

#### TWITTER COUNTS

In [11]:
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

In [12]:
# This is the resulting data from twitter_api.py.

# Downloading it from the link on the Project Details page.

lines=[]
with open('tweet-json.txt','rt') as file:
    for line in file:
        lines.append(line)

In [13]:
# Extracting required data from each line.
f=[]
for line in lines:
    g=[]
    k=line.find('id_str')
    s=line[k+10:k+28]
    g.append(s)
    k=line.find("retweet_count")
    s=line[k+16:k+20]
    g.append(s)
    k=line.find('favorite_count')
    s=line[k+17:k+22]
    g.append(s)
    f.append(g)

In [14]:
# Converting the above extracted data into a pandas dataframe.
api_data=pd.DataFrame(f,columns=['tweet_id','retweet_count','favorite_count'])

In [15]:
api_data.head()

tweet_id retweet_count favorite_count
0  892420643555336193          8853          39467
1  892177421306343426          6514          33819
2  891815181378084864          4328          25461
3  891689557279858688          8964          42908
4  891327558926688256          9774          41048

#### ASSESSING 

In [16]:
# View and assess data
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [17]:
image_predictions.sample(5)

tweet_id                                            jpg_url  \
1462  778383385161035776    https://pbs.twimg.com/media/Cs1fjyqWIAE2jop.jpg   
614   680161097740095489    https://pbs.twimg.com/media/CXBq6RPWkAAaNuU.jpg   
991   708026248782585858  https://pbs.twimg.com/ext_tw_video_thumb/70802...   
146   668627278264475648    https://pbs.twimg.com/media/CUdw9thWsAA4mB9.jpg   
1877  846042936437604353    https://pbs.twimg.com/media/C72_iaUVUAEhZSn.jpg   

      img_num                p1   p1_conf  p1_dog                        p2  \
1462        1            collie  0.345266    True                    borzoi   
614         1          bluetick  0.268681    True        miniature_pinscher   
991         1          malinois  0.786468    True  Chesapeake_Bay_retriever   
146         1    French_bulldog  0.965403    True                       pug   
1877        1  golden_retriever  0.961110    True        Labrador_retriever   

       p2_conf  p2_dog               p3   p3_conf  p3_dog  
1462  0.312823    True    Border_collie  0.213011    True  
614   0.125652    True   English_setter  0.089373    True  
991   0.068979    True      Siamese_cat  0.029304   False  
146   0.008604    True      Boston_bull  0.008004    True  
1877  0.016695    True  Tibetan_mastiff  0.009082    True

In [18]:
# Get number of columns and rows
image_predictions.shape

(2075, 12)

In [19]:
# Get sum of duplicated entries
sum(image_predictions.duplicated())

0

In [20]:
image_predictions.isnull().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

In [21]:
image_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
Chesapeake_Bay_retriever      23
miniature_pinscher            23
seat_belt                     22
Siberian_husky                20
Staffordshire_bullterrier     20
German_shepherd               20
Cardigan                      19
web_site                      19
Shetland_sheepdog             18
Maltese_dog                   18
Eskimo_dog                    18
beagle                        18
teddy                         18
Shih-Tzu                      17
Lakeland_terrier              17
Rottweiler                    17
kuvasz                        16
Italian_greyhound             16
          

In [22]:
image_predictions.p2.value_counts()

Labrador_retriever                104
golden_retriever                   92
Cardigan                           73
Chihuahua                          44
Pomeranian                         42
French_bulldog                     41
Chesapeake_Bay_retriever           41
toy_poodle                         37
cocker_spaniel                     34
miniature_poodle                   33
Siberian_husky                     33
beagle                             28
Pembroke                           27
Eskimo_dog                         27
collie                             27
kuvasz                             26
Italian_greyhound                  22
Pekinese                           21
American_Staffordshire_terrier     21
malinois                           20
toy_terrier                        20
Samoyed                            20
chow                               20
miniature_pinscher                 20
Boston_bull                        19
Norwegian_elkhound                 19
Staffordshir

In [23]:
image_predictions.p3.value_counts()

Labrador_retriever                79
Chihuahua                         58
golden_retriever                  48
Eskimo_dog                        38
kelpie                            35
kuvasz                            34
Staffordshire_bullterrier         32
chow                              32
cocker_spaniel                    31
beagle                            31
toy_poodle                        29
Pomeranian                        29
Pekinese                          29
Great_Pyrenees                    27
Pembroke                          27
Chesapeake_Bay_retriever          27
French_bulldog                    26
malamute                          26
American_Staffordshire_terrier    24
pug                               23
Cardigan                          23
basenji                           21
bull_mastiff                      20
toy_terrier                       20
Siberian_husky                    19
Shetland_sheepdog                 17
Boston_bull                       17
d

In [24]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [25]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [26]:
image_predictions.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

### END

In [27]:
# View and assess data
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [28]:
# Get table shape
twitter_archive.shape

(2356, 17)

In [29]:
# Get summary info
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [30]:
twitter_archive.name.value_counts()

None        745
a            55
Charlie      12
Oliver       11
Lucy         11
Cooper       11
Lola         10
Penny        10
Tucker       10
Winston       9
Bo            9
Sadie         8
the           8
Daisy         7
Bailey        7
Toby          7
Buddy         7
an            7
Jax           6
Leo           6
Scout         6
Rusty         6
Oscar         6
Koda          6
Jack          6
Dave          6
Stanley       6
Milo          6
Bella         6
Gus           5
           ... 
Ito           1
Schnozz       1
Livvie        1
Leonidas      1
Olaf          1
light         1
Sparky        1
Jimbo         1
Harlso        1
Lugan         1
Bayley        1
Lassie        1
Buckley       1
Bode          1
Ivar          1
Willem        1
Amy           1
Amber         1
Milky         1
Tilly         1
Sweet         1
Jeffrie       1
Mack          1
Danny         1
Tess          1
Raphael       1
Eugene        1
Travis        1
Simba         1
Luther        1
Name: name, Length: 957,

In [31]:
# Get summary statistics
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [32]:
# Find null values
twitter_archive.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [33]:
# Get duplicated count
sum(twitter_archive.duplicated())

0

In [34]:
# Lets check if there are any denominators greater or less than 10
not_10_denom = twitter_archive.query('rating_denominator != 10')
not_10_denom

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
342   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
1663  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                             

In [35]:
# Lets get the count of not_10_denom
not_10_denom.count()[0]

23

In [36]:
twitter_archive.rating_numerator.value_counts().head()

12    558
11    464
10    461
13    351
9     158
Name: rating_numerator, dtype: int64

In [37]:
# Lets look for the highest numerator
twitter_archive.rating_numerator.max()

1776

In [38]:
# Lets look for lowest numerator
twitter_archive.rating_numerator.min()

0

### END

In [39]:
# View and assess data
api_data.head(1)

tweet_id retweet_count favorite_count
0  892420643555336193          8853          39467

In [40]:
api_data.tail()

tweet_id retweet_count favorite_count
2349  666049248165822465          41,           111, 
2350  666044226329800704          147,          311, 
2351  666033412701032449          47,           128, 
2352  666029285002620928          48,           132, 
2353  666020888022790149          532,          2535,

In [41]:
# Get table info summary
api_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null object
retweet_count     2354 non-null object
favorite_count    2354 non-null object
dtypes: object(3)
memory usage: 55.2+ KB


In [42]:
#check for null values
api_data.isnull().sum()

tweet_id          0
retweet_count     0
favorite_count    0
dtype: int64

In [43]:
api_data.shape

(2354, 3)

In [44]:
# Check for duplicate column names in all tables
columns = pd.Series(list(twitter_archive) + 
    list(image_predictions) + list(api_data))
columns[columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object

In [45]:
row_count_archive = twitter_archive.shape[0]

In [46]:
row_count_image = image_predictions.shape[0]

### END

QUALITY

image_predictions table

- p1, p2, p3 table contents are not uniformly in lower or upper case
- Duplicate images
- some dog breed predictions are other animals eg hen, ox, shopping_cart, desktop_computer, guinea_pig, orange, papar_towel, convertible, bagel, cock, desk, otter, skunk, sports_car, barracouta, sea_lion, tow_truck, indian_elephant, ice_lolly, fur_coat, bookcase, great_grey_owl, hamster, gar, car_wheel, can_opener, banana, spatula. From value table all are 1
- tweet_id is integer should be string- same for other tables

twitter_archive
- missing values in columns. see drop columns for list
- timestamp is an object not timestamp
- Entries in name columns are begin with upper case and lower case
- some names not provided or not names
- Denominators that are not 10
- Columns doggo, floofer, pupper, and puppo have None as entries which should be null
- tweet_id should be string.

api_data
- columns are strings instead of integer
- some entries have comma after them

TIDINESS

image_predictions table

- prediction, probability and dog and be merged or done another wayColumn name for each type of dog. Should be 1 column and 4 categorical choices
- change column names

twitter_archive

- drop columns - in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp

api_data

- merge api data with twitter archive and image on id
- api has 2354 rows while archive had 2356 while image has 2075. Merge
- find 2 missing rows in api

- Different columns for every dog status - make categorical

### CLEANING

In [47]:
# Make copies of all dataframes
image_clean = image_predictions.copy()
archive_clean = twitter_archive.copy()
api_clean = api_data.copy()

### image_clean

QUALITY

image_predictions table

- p1, p2, p3 table contents are not uniformly in lower or upper case
- Duplicate images
- Change column names for p1, p2, p3 to more descriptive name
- some dog breed predictions are other animals eg hen, ox, shopping_cart, desktop_computer, guinea_pig, orange, papar_towel, convertible, bagel, cock, desk, otter, skunk, sports_car, barracouta, sea_lion, tow_truck, indian_elephant, ice_lolly, fur_coat, bookcase, great_grey_owl, hamster, gar, car_wheel, can_opener, banana, spatula. From value table all are 1
- tweet_id is integer should be string- same for other tables
- drop p2, p3 and associated columns because p1 has the highest prediction mean
- p1 is an object instead of categorical


In [48]:
image_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

**Define:**

Capitalize dog breed name in P1, P2 and P3

**Code**

In [49]:
# Capitalize breed names in P1
image_clean.p1 = image_clean.p1.str.capitalize()

In [50]:
# Capitalize breed names in P2
image_clean.p2 = image_clean.p2.str.capitalize()

In [51]:
# Capitalize breed names in P3
image_clean.p3 = image_clean.p2.str.capitalize()

**Test**

In [52]:
image_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              Collie   
1        1                 Redbone  0.506826    True  Miniature_pinscher   
2        1         German_shepherd  0.596461    True            Malinois   
3        1     Rhodesian_ridgeback  0.408143    True             Redbone   
4        1      Miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                  p3   p3_conf  p3_dog  
0  0.156665    True              Collie  0.061428    True  
1  0.074192    True  Miniature_pinscher  0.072010    True  
2  0.138584    True            Malinois  0.116197    True  
3  0.360687    True             Redbone  0.222752    True  
4  0.243682    True          Rottweiler  0.154629    True

**Define**

Change tweet_id from interger to string using astype

**Code**

In [53]:
image_clean.tweet_id = image_clean.tweet_id.astype(str)

**Test**

In [54]:
image_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


**Define**

Use rename to change p1 column name

**Code**

In [55]:
image_clean.rename(columns = {'img_num': 'nof_images', 
                               'p1': 'best_prediction', 
                               'p1_conf': 'confidence', 
                               'p1_dog': 'dog_or_not'}, inplace = True)

**Test**

In [56]:
image_clean.head(1)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   

   nof_images         best_prediction  confidence  dog_or_not      p2  \
0           1  Welsh_springer_spaniel    0.465074        True  Collie   

    p2_conf  p2_dog      p3   p3_conf  p3_dog  
0  0.156665    True  Collie  0.061428    True

In [57]:
image_predictions.p3_conf.mean()

0.060324168618106515

**Define**
Drop columns p2, p2_conf, p2_dog, p3, p3_conf, p3_dog using drop function

**Code**

In [58]:
image_clean.drop(['p2', 'p2_conf', 'p2_dog', 'p3', 
                  'p3_conf', 'p3_dog'], inplace = True, axis = 1)

**Test**

In [59]:
image_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 6 columns):
tweet_id           2075 non-null object
jpg_url            2075 non-null object
nof_images         2075 non-null int64
best_prediction    2075 non-null object
confidence         2075 non-null float64
dog_or_not         2075 non-null bool
dtypes: bool(1), float64(1), int64(1), object(3)
memory usage: 83.2+ KB


**Define**

Change the best prediction column to categorical using astype

**Code**

In [60]:
image_clean.best_prediction = image_clean.best_prediction.astype('category')

**Test**

In [61]:
image_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 6 columns):
tweet_id           2075 non-null object
jpg_url            2075 non-null object
nof_images         2075 non-null int64
best_prediction    2075 non-null category
confidence         2075 non-null float64
dog_or_not         2075 non-null bool
dtypes: bool(1), category(1), float64(1), int64(1), object(2)
memory usage: 84.0+ KB


In [62]:
image_clean.best_prediction.value_counts()

Golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
Pug                           57
Chow                          44
Samoyed                       43
Toy_poodle                    39
Pomeranian                    38
Cocker_spaniel                30
Malamute                      30
French_bulldog                26
Miniature_pinscher            23
Chesapeake_bay_retriever      23
Seat_belt                     22
Siberian_husky                20
German_shepherd               20
Staffordshire_bullterrier     20
Cardigan                      19
Web_site                      19
Teddy                         18
Eskimo_dog                    18
Maltese_dog                   18
Shetland_sheepdog             18
Beagle                        18
Lakeland_terrier              17
Rottweiler                    17
Shih-tzu                      17
Kuvasz                        16
Italian_greyhound             16
          

**Define**

Create data frame for dogs only using query

**Code**

In [63]:
dogs_only = image_clean.query('dog_or_not == True')
dogs_only.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   nof_images         best_prediction  confidence  dog_or_not  
0           1  Welsh_springer_spaniel    0.465074        True  
1           1                 Redbone    0.506826        True  
2           1         German_shepherd    0.596461        True

**Test**

In [64]:
dogs_only.best_prediction.value_counts().head(8)

Golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
Pug                    57
Chow                   44
Samoyed                43
Toy_poodle             39
Name: best_prediction, dtype: int64

#### TIDINESS

api_data

merge api data with twitter archive and image on id
api has 2354 rows while archive had 2356 while image has 2075. Merge
find 2 missing rows in api

- Different columns for every dog status - make categorical

#### END

### archive_clean

QUALITY

twitter_archive

- tweet_id should be string.
- timestamp is an object not timestamp
- In name column some names have None while others are words like a, an, the
- Denominators that are not 10 and outliers in numerators


TIDINESS
- Columns doggo, floofer, pupper, and puppo should be one column instead of 4 
- Columns that are irrelevant in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp


**Define**

Change tweet_id column from integer to string using astype


**Code**

In [65]:
archive_clean.tweet_id = archive_clean.tweet_id.astype(str)

**Test**

In [66]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), o

**Define**

Convert timestamp column from object type to timestamp using

**Code**

In [67]:
archive_clean.timestamp =  pd.to_datetime(archive_clean['timestamp'])

**Test**

In [68]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datetime64[ns](

In [69]:
archive_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

**Define**

- Replace the word None with NaN
- Make a list of most common prepositions such as a, an, and, the and replace them with NaN

**Code**

In [70]:
# Replace None with NaN
archive_clean.name = archive_clean.name.replace('None','NaN')

In [71]:
# Create list of most common prepositions
prepositions = ['O', 'a', 'about', 'above', 'after', 'again', 
             'against', 'all', 'all', 'am', 'an', 'an', 'and', 
             'any', 'are', 'as', 'at', 'at', 'be', 'because', 
             'been', 'before', 'being', 'below', 'between', 
             'both', 'but', 'by', 'by', 'can', 'did', 'do', 
             'does', 'doing', 'don', 'down', 'during', 'each', 
             'few', 'for', 'from', 'further', 'had', 'has', 
             'have', 'having', 'he', 'her', 'here', 'hers', 
             'herself', 'him', 'himself', 'his', 'how', 'i', 
             'if', 'in', 'into', 'is', 'it', 'its', 'itself', 
             'just', 'just', 'life', 'light', 'me', 'more', 
             'most', 'my', 'my', 'myself', 'no', 'nor', 'not', 
             'not', 'now', 'of', 'off', 'old', 'on', 'once', 
             'only', 'or', 'other', 'our', 'ours', 'ourselves', 
             'out', 'over', 'own', 'quite', 's', 'same', 'she', 
             'should', 'so', 'some', 'space', 'such', 
             't', 'than', 'that', 'the', 'the', 'their', 'theirs', 
             'them', 'themselves', 'then', 'there', 'these', 'they', 
             'this', 'this', 'those', 'through', 'to', 'too', 'under', 
             'until', 'up', 'very', 'very', 'was', 'we', 'were', 
             'what', 'when', 'where', 'which', 'while', 'who', 
             'whom', 'why', 'will', 'with', 'within', 'without', 'you', 'your', 'yours', 
             'yourself', 'yourselves']

In [72]:
# Create a list of prepositions
prepositions = set(prepositions)

In [73]:
# Write to disk to save
with open('prepositions.txt', 'w') as f:
    for word in prepositions:
        f.write(word + '\n')

In [74]:
# Confirm it it works and then sort
with open('prepositions.txt', 'r') as f:
    prepositions = [line.rstrip('\n') for line in f]
prepositions.sort()

In [75]:
# Replace the prepositions with NaNs.
archive_clean.name.replace(prepositions, np.nan, inplace = True)

**Test**

In [76]:
archive_clean.name.value_counts()

NaN         745
Charlie      12
Lucy         11
Oliver       11
Cooper       11
Penny        10
Tucker       10
Lola         10
Bo            9
Winston       9
Sadie         8
Toby          7
Daisy         7
Buddy         7
Bailey        7
Milo          6
Bella         6
Dave          6
Jax           6
Koda          6
Jack          6
Leo           6
Scout         6
Rusty         6
Oscar         6
Stanley       6
Sunny         5
George        5
Bentley       5
Phil          5
           ... 
Blipson       1
Aldrick       1
Crimson       1
Devón         1
Schnozz       1
Livvie        1
Leonidas      1
Olaf          1
Ito           1
Lugan         1
Lassie        1
Simba         1
Milky         1
Buckley       1
Bode          1
Ivar          1
Willem        1
Amy           1
Amber         1
Jareld        1
Tilly         1
Travis        1
Sweet         1
Jeffrie       1
Mack          1
Danny         1
Tess          1
Raphael       1
Eugene        1
Ulysses       1
Name: name, Length: 939,

In [77]:
#archive_clean['rating'] = archive_clean.rating_numerator / archive_clean.rating_denominator

In [78]:
archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   
1 2017-08-01 00:17:27  <a href="http://twitter.com/download/iphone" r...   
2 2017-07-31 00:18:03  <a href="http://twitter.com/download/iphone" r...   
3 2017-07-30 15:58:51  <a href="http://twitter.com/download/iphone" r...   
4 2017-07-29 16:00:24  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [79]:
archive_clean.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

**Define**

- Use query to allow only for denominators should be 10 and numerators should be less than 16

**Code**

In [80]:
archive_clean = archive_clean.query('rating_numerator < 16 and rating_denominator == 10')

**Test**

In [81]:
archive_clean.rating_denominator.value_counts()

10    2323
Name: rating_denominator, dtype: int64

In [82]:
archive_clean.rating_numerator.value_counts()

12    558
11    463
10    461
13    351
9     156
8     102
7      54
14     54
5      37
6      32
3      19
4      15
2       9
1       8
15      2
0       2
Name: rating_numerator, dtype: int64

**Define**

- Drops columns that are not relevant in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp

**Code**

In [83]:
# Drop columns
archive_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace = True, axis = 1)

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


**Test**

In [84]:
archive_clean.head(0)

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

**Define**

Combine columns doggo, floofer, pupper, and puppo should be one column called dog_stage

**Code**

In [85]:
dog_stage = ['doggo', 'floofer', 'pupper', 'puppo']
def my_dog_stage(stages):
    for stage in stages:
        if stage in dog_stage:
            return stage
        else:
            pass
    return stage

In [86]:
# Combine all dog stages into 1 column 
archive_clean['dog_stage'] = archive_clean[['doggo', 'floofer', 
    'pupper', 'puppo']].apply(my_dog_stage, axis = 1)


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [87]:
# Drop columns not needed
archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], 
    inplace = True, axis = 1)
archive_clean.dog_stage = archive_clean.dog_stage.astype('category')

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)
/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [88]:
# Convert data type to category 

archive_clean.dog_stage = archive_clean.dog_stage.astype('category')

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [89]:
# Capitalize stages in dog_stage
archive_clean.dog_stage = archive_clean.dog_stage.str.capitalize()

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


**Test**

In [90]:
archive_clean

tweet_id           timestamp  \
0     892420643555336193 2017-08-01 16:23:56   
1     892177421306343426 2017-08-01 00:17:27   
2     891815181378084864 2017-07-31 00:18:03   
3     891689557279858688 2017-07-30 15:58:51   
4     891327558926688256 2017-07-29 16:00:24   
5     891087950875897856 2017-07-29 00:08:17   
6     890971913173991426 2017-07-28 16:27:12   
7     890729181411237888 2017-07-28 00:22:40   
8     890609185150312448 2017-07-27 16:25:51   
9     890240255349198849 2017-07-26 15:59:51   
10    890006608113172480 2017-07-26 00:31:25   
11    889880896479866881 2017-07-25 16:11:53   
12    889665388333682689 2017-07-25 01:55:32   
13    889638837579907072 2017-07-25 00:10:02   
14    889531135344209921 2017-07-24 17:02:04   
15    889278841981685760 2017-07-24 00:19:32   
16    888917238123831296 2017-07-23 00:22:39   
17    888804989199671297 2017-07-22 16:56:37   
18    888554962724278272 2017-07-22 00:23:06   
19    888202515573088257 2017-07-21 01:02:36   
20    888078434458587136 2017-07-20 16:49:33   
21    887705289381826560 2017-07-19 16:06:48   
22    887517139158093824 2017-07-19 03:39:09   
23    887473957103951883 2017-07-19 00:47:34   
24    887343217045368832 2017-07-18 16:08:03   
25    887101392804085760 2017-07-18 00:07:08   
26    886983233522544640 2017-07-17 16:17:36   
27    886736880519319552 2017-07-16 23:58:41   
28    886680336477933568 2017-07-16 20:14:00   
29    886366144734445568 2017-07-15 23:25:31   
...                  ...                 ...   
2325  666418789513326592 2015-11-17 00:53:15   
2326  666411507551481857 2015-11-17 00:24:19   
2327  666407126856765440 2015-11-17 00:06:54   
2328  666396247373291520 2015-11-16 23:23:41   
2329  666373753744588802 2015-11-16 21:54:18   
2330  666362758909284353 2015-11-16 21:10:36   
2331  666353288456101888 2015-11-16 20:32:58   
2332  666345417576210432 2015-11-16 20:01:42   
2333  666337882303524864 2015-11-16 19:31:45   
2334  666293911632134144 2015-11-16 16:37:02   
2336  666273097616637952 2015-11-16 15:14:19   
2337  666268910803644416 2015-11-16 14:57:41   
2338  666104133288665088 2015-11-16 04:02:55   
2339  666102155909144576 2015-11-16 03:55:04   
2340  666099513787052032 2015-11-16 03:44:34   
2341  666094000022159362 2015-11-16 03:22:39   
2342  666082916733198337 2015-11-16 02:38:37   
2343  666073100786774016 2015-11-16 01:59:36   
2344  666071193221509120 2015-11-16 01:52:02   
2345  666063827256086533 2015-11-16 01:22:45   
2346  666058600524156928 2015-11-16 01:01:59   
2347  666057090499244032 2015-11-16 00:55:59   
2348  666055525042405380 2015-11-16 00:49:46   
2349  666051853826850816 2015-11-16 00:35:11   
2350  666050758794694657 2015-11-16 00:30:50   
2351  666049248165822465 2015-11-16 00:24:50   
2352  666044226329800704 2015-11-16 00:04:52   
2353  666033412701032449 2015-11-15 23:21:54   
2354  666029285002620928 2015-11-15 23:05:30   
2355  666020888022790149 2015-11-15 22:32:08   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="http://twitter.com/download/iphone" r...   
10    <a href="http://twitter.com/download/iphone" r...   
11    <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
13    <a href="http://twitter.com/download/iphone" r...   
14    <a href="http://twitter.com/download/iphone" r...   
15    <a href="http://twitter.com/download/iphone" r...   
16    <a href="http://twitter.com/do

**Define**

Replace None with NaN in dog_stage column

**Code**

In [91]:
# Replace None with NaNs.
archive_clean.dog_stage = archive_clean.dog_stage.replace('None','NaN')

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


**Test**

In [92]:
archive_clean

tweet_id           timestamp  \
0     892420643555336193 2017-08-01 16:23:56   
1     892177421306343426 2017-08-01 00:17:27   
2     891815181378084864 2017-07-31 00:18:03   
3     891689557279858688 2017-07-30 15:58:51   
4     891327558926688256 2017-07-29 16:00:24   
5     891087950875897856 2017-07-29 00:08:17   
6     890971913173991426 2017-07-28 16:27:12   
7     890729181411237888 2017-07-28 00:22:40   
8     890609185150312448 2017-07-27 16:25:51   
9     890240255349198849 2017-07-26 15:59:51   
10    890006608113172480 2017-07-26 00:31:25   
11    889880896479866881 2017-07-25 16:11:53   
12    889665388333682689 2017-07-25 01:55:32   
13    889638837579907072 2017-07-25 00:10:02   
14    889531135344209921 2017-07-24 17:02:04   
15    889278841981685760 2017-07-24 00:19:32   
16    888917238123831296 2017-07-23 00:22:39   
17    888804989199671297 2017-07-22 16:56:37   
18    888554962724278272 2017-07-22 00:23:06   
19    888202515573088257 2017-07-21 01:02:36   
20    888078434458587136 2017-07-20 16:49:33   
21    887705289381826560 2017-07-19 16:06:48   
22    887517139158093824 2017-07-19 03:39:09   
23    887473957103951883 2017-07-19 00:47:34   
24    887343217045368832 2017-07-18 16:08:03   
25    887101392804085760 2017-07-18 00:07:08   
26    886983233522544640 2017-07-17 16:17:36   
27    886736880519319552 2017-07-16 23:58:41   
28    886680336477933568 2017-07-16 20:14:00   
29    886366144734445568 2017-07-15 23:25:31   
...                  ...                 ...   
2325  666418789513326592 2015-11-17 00:53:15   
2326  666411507551481857 2015-11-17 00:24:19   
2327  666407126856765440 2015-11-17 00:06:54   
2328  666396247373291520 2015-11-16 23:23:41   
2329  666373753744588802 2015-11-16 21:54:18   
2330  666362758909284353 2015-11-16 21:10:36   
2331  666353288456101888 2015-11-16 20:32:58   
2332  666345417576210432 2015-11-16 20:01:42   
2333  666337882303524864 2015-11-16 19:31:45   
2334  666293911632134144 2015-11-16 16:37:02   
2336  666273097616637952 2015-11-16 15:14:19   
2337  666268910803644416 2015-11-16 14:57:41   
2338  666104133288665088 2015-11-16 04:02:55   
2339  666102155909144576 2015-11-16 03:55:04   
2340  666099513787052032 2015-11-16 03:44:34   
2341  666094000022159362 2015-11-16 03:22:39   
2342  666082916733198337 2015-11-16 02:38:37   
2343  666073100786774016 2015-11-16 01:59:36   
2344  666071193221509120 2015-11-16 01:52:02   
2345  666063827256086533 2015-11-16 01:22:45   
2346  666058600524156928 2015-11-16 01:01:59   
2347  666057090499244032 2015-11-16 00:55:59   
2348  666055525042405380 2015-11-16 00:49:46   
2349  666051853826850816 2015-11-16 00:35:11   
2350  666050758794694657 2015-11-16 00:30:50   
2351  666049248165822465 2015-11-16 00:24:50   
2352  666044226329800704 2015-11-16 00:04:52   
2353  666033412701032449 2015-11-15 23:21:54   
2354  666029285002620928 2015-11-15 23:05:30   
2355  666020888022790149 2015-11-15 22:32:08   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="http://twitter.com/download/iphone" r...   
10    <a href="http://twitter.com/download/iphone" r...   
11    <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
13    <a href="http://twitter.com/download/iphone" r...   
14    <a href="http://twitter.com/download/iphone" r...   
15    <a href="http://twitter.com/download/iphone" r...   
16    <a href="http://twitter.com/do

#### END

### api_data

QUALITY

api_data

- Columns retweet_count and favorite_count are object instead of integers
- Some entries in retweet_count and favorite_count have comma after them and quotation marks.


TIDINESS

**Define**

- Extract numerical values from the affected columns

**Code**

In [93]:
# Extract integers from retweet count column
api_clean.retweet_count = api_clean.retweet_count.str.extract('(\d+)')

In [94]:
# Extract integers from favorite count column
api_clean.favorite_count = api_clean.favorite_count.str.extract('(\d+)')

**Test**

In [95]:
api_clean

tweet_id retweet_count favorite_count
0     892420643555336193          8853          39467
1     892177421306343426          6514          33819
2     891815181378084864          4328          25461
3     891689557279858688          8964          42908
4     891327558926688256          9774          41048
5     891087950875897856          3261          20562
6     890971913173991426          2158          12041
7     890729181411237888          1671          56848
8     890609185150312448          4429          28226
9     890240255349198849          7711          32467
10    890006608113172480          7624          31166
11    889880896479866881          5156          28268
12    889665388333682689          8538          38818
13    889638837579907072          4735          27672
14    889531135344209921          2321          15359
15    889278841981685760          5637          25652
16    888917238123831296          4709          29611
17    888804989199671297          4559          26080
18    888554962724278272          3732          20290
19    888078434458587136          3653          22201
20    887705289381826560          5609          30779
21    887517139158093824          1208          46959
22    887473957103951883          1878          69871
23    887343217045368832          1073          34222
24    887101392804085760          6167          31061
25    886983233522544640          8084          35859
26    886736880519319552          3443          12306
27    886680336477933568          4610          22798
28    886366144734445568          3316          21524
29    886267009285017600             4            117
...                  ...           ...            ...
2324  666411507551481857           339            459
2325  666407126856765440            44            113
2326  666396247373291520            92            172
2327  666373753744588802           100            194
2328  666362758909284353           595            804
2329  666353288456101888            77            229
2330  666345417576210432           146            307
2331  666337882303524864            96            204
2332  666293911632134144           368            522
2333  666287406224695296            71            152
2334  666273097616637952            82            184
2335  666268910803644416            37            108
2336  666104133288665088          6871          14765
2337  666102155909144576            16             81
2338  666099513787052032            73            164
2339  666094000022159362            79            169
2340  666082916733198337            47            121
2341  666073100786774016           174            335
2342  666071193221509120            67            154
2343  666063827256086533           232            496
2344  666058600524156928            61            115
2345  666057090499244032           146            304
2346  666055525042405380           261            448
2347  666051853826850816           879           1253
2348  666050758794694657            60            136
2349  666049248165822465            41            111
2350  666044226329800704           147            311
2351  666033412701032449            47            128
2352  666029285002620928            48            132
2353  666020888022790149           532           2535

[2354 rows x 3 columns]

**Define**

Convert retweet and favorite count data type to integer

**Code**

In [96]:
api_clean[["retweet_count", "favorite_count"]] = api_clean[["retweet_count", "favorite_count"]].apply(pd.to_numeric)

**Test**

In [97]:
api_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null object
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(2), object(1)
memory usage: 55.2+ KB
